src/python/pants/backend/terraform/tool.py

```diff
@@ -2,44 +2,165 @@
+
+        return tuple(OrderedSet(iter_path_entries()))
+
+
+@dataclass(frozen=True)
+class TerraformSetup:
+    path: BinaryPath
+
+
+@rule
+async def find_terraform(terraform: TerraformSubsystem) -> TerraformSetup:
+    env = await Get(Environment, EnvironmentRequest(["PATH"]))
+    search_paths = terraform.search_paths(env)
+
+    all_terraform_binary_paths = await Get(
+        BinaryPaths,
+        BinaryPathRequest(
+            search_path=search_paths,
+            binary_name="terraform",
+            test=BinaryPathTest(["version"]),
+        ),
     )
-    default_url_platform_mapping = {
-        "macos_arm64": "darwin_amd64",
-        "macos_x86_64": "darwin_amd64",
-        "linux_x86_64": "linux_amd64",
-    }
 
-    @classproperty
-    def default_known_versions(cls):
-        return [
-            "1.0.7|macos_arm64 |cbab9aca5bc4e604565697355eed185bb699733811374761b92000cc188a7725|32071346",
-            "1.0.7|macos_x86_64|80ae021d6143c7f7cbf4571f65595d154561a2a25fd934b7a8ccc1ebf3014b9b|33020029",
-            "1.0.7|linux_x86_64|bc79e47649e2529049a356f9e60e06b47462bf6743534a10a4c16594f443be7b|32671441",
-        ]
+    if not all_terraform_binary_paths.paths:
+        raise BinaryNotFoundError(
+            "Cannot find any `terraform` binaries using the option "
+            f"`[terraform].search_paths`: {list(search_paths)}\n\n"
+            "To fix, please install Terraform and ensure that the `terraform` binary "
+            "is discoverable via `[terraform].search_paths`."
+        )
+
+    version_results = await MultiGet(
+        Get(
+            ProcessResult,
+            Process(
+                (binary_path.path, "version", "-json"),
+                description=f"Determine Terraform version for {binary_path.path}",
+                level=LogLevel.DEBUG,
+                cache_scope=ProcessCacheScope.PER_RESTART_SUCCESSFUL,
+            ),
+        )
+        for binary_path in all_terraform_binary_paths.paths
+    )
+
+    constraints_set = terraform.version_constraint_set
+
+    invalid_versions = []
+    for binary_path, version_result in zip(all_terraform_binary_paths.paths, version_results):
+        try:
+            version = Version(
+                json.loads(version_result.stdout.decode("utf-8"))["terraform_version"]
+            )
+        except InvalidVersion:
+            raise AssertionError(
+                f"Failed to parse `terraform version` output for {binary_path}. Please open an issue at "
+                f"https://github.com/pantsbuild/pants/issues/new/choose with the below data."
+                f"\n\n"
+                f"{version_result}"
+            )
+
+        if constraints_set.contains(version):
+            return TerraformSetup(binary_path)
+
+        invalid_versions.append((binary_path.path, version))
+
+    bulleted_list_sep = "\n  * "
+    invalid_versions_str = bulleted_list_sep.join(
+        f"{path}: {version}" for path, version in sorted(invalid_versions)
+    )
```

Comment:
This should use `bullet_list` from `pants.util.strutil`, landed in the past day. Check `golang.py` for an example